## Coursera Capstone

This notebook will be mainly used for the capstone project!

In [2]:
import pandas as pd
import numpy as np

print ('Hello Capstone Project Course!')

Hello Capstone Project Course!


### Introduction/Business Problem

The manager of a chain of niche bakeries has been in touch as the company would like to expand to Scotland, having opened shops across 10 locations across England. They have requested that I compare the postcodes of the Scottish capital, Edinburgh, to inform what location would be best suited for their business. Of particular interest are the existing cafés, bistros and bakeries that would, naturally, compete with their branch for customers, as well as the general layout of the city. The client's key requirement is that the bakery should be centrally located, to maximise foot traffic during busy times such as festivals.

### Data

I will use the Foursquare API (https://foursquare.com/) for this analysis:
- The location of interest is Edinburgh.
- Using the latitude and longitude of each postcode in the city, I will explore the city centre, paying specific attention to eateries (cafés, bistros and bakeries). This will involve the use of the explore function and a k-means clustering algorithm.
- Finally, I will use the Folium library to map out the city, showing the clusters of eateries in Edinburgh.

To obtain coordinates for each location, I will use the '2020-2 Scottish Postcode Directory Files' datasets provided by National Records Scotland. This is a list of active and deleted postcodes in Scotland, offered freely online (https://www.nrscotland.gov.uk/statistics-and-data/geography/nrs-postcode-extract).


# Analysis


To download the required libraries:


In [3]:
import numpy as np # to handle vectors

import pandas as pd # for analysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # to convert an address into latitude and longitude

import requests # to handle requests
from pandas.io.json import json_normalize # to tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    openssl-1.1.1g             |       h516909a_1         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

### 1. Explore Dataset

To import the dataset containing coordinates of different postcodes in Scotland:

In [4]:

import types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

df_data_1 = pd.read_csv(body)

df_data_1.head(20)


/opt/conda/envs/Python36/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (46) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Postcode PostcodeDistrict PostcodeSector  DateOfIntroduction  \
0    AB1 0AA              AB1          AB1 0   1/1/1980 00:00:00   
1    AB1 0AB              AB1          AB1 0   1/8/1973 00:00:00   
2    AB1 0AD              AB1          AB1 0   1/8/1973 00:00:00   
3    AB1 0AE              AB1          AB1 0   1/2/1994 00:00:00   
4    AB1 0AJ              AB1          AB1 0   1/8/1973 00:00:00   
5    AB1 0AL              AB1          AB1 0   1/8/1973 00:00:00   
6    AB1 0AP              AB1          AB1 0   1/8/1973 00:00:00   
7    AB1 0AQ              AB1          AB1 0   1/8/1973 00:00:00   
8    AB1 0AR              AB1          AB1 0   1/8/1973 00:00:00   
9    AB1 0AS              AB1          AB1 0   1/8/1973 00:00:00   
10   AB1 0AT              AB1          AB1 0   1/8/1973 00:00:00   
11  AB42 0HJ             AB42         AB42 0  12/7/1996 00:00:00   
12  AB42 0HL             AB42         AB42 0   1/4/1996 00:00:00   
13  AB42 0HP             AB42         AB42 0  12/7/1996 00:00:00   
14  AB42 0HW             AB42         AB42 0  12/7/1996 00:00:00   
15  AB42 0HX             AB42         AB42 0   1/4/1996 00:00:00   
16  AB42 0HY             AB42         AB42 0  12/7/1996 00:00:00   
17  AB42 0HZ             AB42         AB42 0  12/7/1996 00:00:00   
18  AB42 0LE             AB42         AB42 0   1/4/1996 00:00:00   
19  AB42 0LT             AB42         AB42 0   1/4/1996 00:00:00   

       DateOfDeletion  GridReferenceEasting  GridReferenceNorthing   Latitude  \
0   1/6/1996 00:00:00                385386                 801193  57.101482   
1   1/4/1996 00:00:00                384939                 801420  57.103507   
2   1/4/1996 00:00:00                384939                 800980  57.099555   
3   1/4/1996 00:00:00                384599                 799300  57.084452   
4   1/4/1996 00:00:00                384739                 801010  57.099817   
5   1/4/1996 00:00:00                384739                 801260  57.102063   
6   1/4/1996 00:00:00                385019                 800740  57.097401   
7   1/4/1996 00:00:00                384799                 800870  57.098562   
8   1/4/1996 00:00:00                386539                 800040  57.091160   
9   1/4/1996 00:00:00                386239                 799140  57.083067   
10  1/4/1996 00:00:00                385639                 799840  57.089336   
11                NaN                409340                 836066  57.414884   
12                NaN                405282                 837376  57.426715   
13                NaN                409022                 836454  57.418375   
14                NaN                409146                 836869  57.422100   
15                NaN                405282                 837044  57.423733   
16                NaN                404941                 837225  57.425362   
17                NaN                405312                 837026  57.423571   
18                NaN                406658                 837160  57.424757   
19                NaN                401715                 836191  57.416097   

    Longitude SplitIndicator CouncilArea2019Code  \
0   -2.242872              N           S12000033   
1   -2.250264              N           S12000033   
2   -2.250237              N           S12000033   
3   -2.255745              N           S12000034   
4   -2.253540              N           S12000033   
5   -2.253555              N           S12000033   
6   -2.248902              N           S12000033   
7   -2.252541              N           S12000033   
8   -2.223778              N           S12000034   
9   -2.228679              N           S12000034   
10  -2.238618              N           S12000034   
11  -1.846190              N           S12000034   
12  -1.913715              N           S12000034   
13  -1.851470              N           S12000034   
14  -1.849390              N           S12000034   
15  -1.913722              N           S1

In [5]:
live_df = df_data_1[df_data_1['DateOfDeletion'].isnull()]  # to eliminate 'dead' postcodes from the dataset

postcodes = pd.DataFrame()
postcodes ['Postcode'] = live_df ['Postcode']
postcodes ['Latitude'] = live_df ['Latitude']
postcodes ['Longitude'] = live_df ['Longitude']
# new dataframe containing only relevant columns

postcodes.reset_index(drop = True, inplace = True)

postcodes.head()


Postcode   Latitude  Longitude
0  AB42 0HJ  57.414884  -1.846190
1  AB42 0HL  57.426715  -1.913715
2  AB42 0HP  57.418375  -1.851470
3  AB42 0HW  57.422100  -1.849390
4  AB42 0HX  57.423733  -1.913722

The dataframe now contains only postcodes that are currently in use in Scotland.

### 2. Define function

To declare a function that will use Foursquare credentials to query the venues in postcodes within Edinburgh:

In [6]:
CLIENT_ID = '' # my Foursquare ID
CLIENT_SECRET = '' # my Foursquare Secret
VERSION = '' # Foursquare API version
LIMIT = 80 # limit of number of venues returned by Foursquare API

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&categoryId=4bf58dd8d48988d1d0941735,4bf58dd8d48988d1e0931735,4bf58dd8d48988d16a941735,52e81612bcbc57f1066b79f1,4bf58dd8d48988d16d941735'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postcode', 
                  'Postcode Latitude', 
                  'Postcode Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

To get the geographical coordinates of Edinburgh:

In [10]:
# Edinburgh
address = 'Edinburgh, UK'

geolocator = Nominatim(user_agent="scot_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Edinburgh are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Edinburgh are 55.9533456, -3.1883749.


Given the client's requirement to site the bakery in the city centre, I limit the dataset to postcodes within a 1km radius of Edinburgh:

In [28]:
from geopy.distance import geodesic

def cleanPostcodes(origin_latitude, origin_longitude, postcodes_original, radius=1000):
    center = (origin_latitude, origin_longitude)

    new_rows = []
    for i, x in postcodes_original.iterrows():
        p = (x['Latitude'], x['Longitude'])
        if geodesic(p, center).meters < radius:
            new_rows.append(x.values)
        
    postcodes_cleaned = pd.DataFrame(new_rows, columns=postcodes_original.columns)
    #.reset_index(drop=True, inplace=True)
    return(postcodes_cleaned)

In [29]:
postcodes_edb = cleanPostcodes(latitude, longitude, postcodes)

In [30]:
postcodes_edb.shape

(667, 3)

This leaves a total of 667 postcodes to be used for Foursquare calls.

In [31]:
edinburgh_venues = getNearbyVenues(names = postcodes_edb['Postcode'],
                                   latitudes = postcodes_edb['Latitude'],
                                   longitudes = postcodes_edb['Longitude']
                                  )

In [32]:
edinburgh_venues.head()

Postcode  Postcode Latitude  Postcode Longitude             Venue  \
0  EH1 1BL          55.950658           -3.191749    artisan gelato   
1  EH1 1BB          55.952027           -3.189565  Cornish Pasty Co   
2  EH1 1BB          55.952027           -3.189565  Millie's Cookies   
3  EH1 1BB          55.952027           -3.189565            Greggs   
4  EH1 1DE          55.950954           -3.189996    artisan gelato   

   Venue Latitude  Venue Longitude Venue Category  
0       55.950581        -3.190650    Gelato Shop  
1       55.952224        -3.189511         Bakery  
2       55.951883        -3.189436         Bakery  
3       55.952774        -3.190335         Bakery  
4       55.950581        -3.190650    Gelato Shop

In [33]:
edinburgh_venues.shape

(479, 7)

Due to overlap within the search radius for each postcode, I have to drop duplicate venues:

In [36]:
edinburgh_venues.drop_duplicates(subset = 'Venue', inplace = True)
edinburgh_venues.head()

Postcode  Postcode Latitude  Postcode Longitude             Venue  \
0  EH1 1BL          55.950658           -3.191749    artisan gelato   
1  EH1 1BB          55.952027           -3.189565  Cornish Pasty Co   
2  EH1 1BB          55.952027           -3.189565  Millie's Cookies   
3  EH1 1BB          55.952027           -3.189565            Greggs   
7  EH1 1AD          55.948907           -3.192608          Demijohn   

   Venue Latitude  Venue Longitude Venue Category  
0       55.950581        -3.190650    Gelato Shop  
1       55.952224        -3.189511         Bakery  
2       55.951883        -3.189436         Bakery  
3       55.952774        -3.190335         Bakery  
7       55.948620        -3.193979   Gourmet Shop

To group venues within each postcode by venue category:

In [37]:
edinburgh_venues.groupby('Postcode').count()

Postcode Latitude  Postcode Longitude  Venue  Venue Latitude  \
Postcode                                                                 
EH1 1AD                   1                   1      1               1   
EH1 1BB                   3                   3      3               3   
EH1 1BL                   1                   1      1               1   
EH1 1BQ                   1                   1      1               1   
EH1 1DR                   3                   3      3               3   
EH1 1EZ                   1                   1      1               1   
EH1 1HR                   2                   2      2               2   
EH1 1JQ                   1                   1      1               1   
EH1 1JX                   1                   1      1               1   
EH1 1LS                   1                   1      1               1   
EH1 1NL                   1                   1      1               1   
EH1 1SX                   1                   1      1               1   
EH1 2HS                   1                   1      1               1   
EH1 2JL                   1                   1      1               1   
EH1 2QP                   1                   1      1               1   
EH1 3AA                   1                   1      1               1   
EH1 3DQ                   1                   1      1               1   
EH1 3HP                   1                   1      1               1   
EH1 3LD                   1                   1      1               1   
EH2 1LN                   1                   1      1               1   
EH2 2BY                   1                   1      1               1   
EH2 2PJ                   1                   1      1               1   
EH2 3AA                   3                   3      3               3   
EH2 3ES                   1                   1      1               1   
EH2 3NS                   1                   1      1               1   
EH3 6LN                   1                   1      1               1   
EH3 6NG                   1                   1      1               1   
EH7 4AA                   1                   1      1               1   
EH7 4AG                   1                   1      1               1   
EH7 4AL                   1                   1      1               1   
EH8 8DF                   1                   1      1               1   
EH8 9AY                   1                   1      1               1   
EH8 9JZ                   1                   1      1               1   

          Venue Longitude  Venue Category  
Postcode                                   
EH1 1AD                 1               1  
EH1 1BB                 3               3  
EH1 1BL                 1               1  
EH1 1BQ                 1               1  
EH1 1DR                 3               3  
EH1 1EZ                 1               1  
EH1 1HR                 2               2  
EH1 1JQ                 1               1  
EH1 1JX                 1               1  
EH1 1LS                 1               1  
EH1 1NL                 1               1  
EH1 1SX                 1               1  
EH1 2HS                 1               1  
EH1 2JL                 1               1  
EH1 2QP                 1               1  
EH1 3AA                 1               1  
EH1 3DQ                 1               1  
EH1 3HP                 1               1  
EH1 3LD                 1               1  
EH2 1LN                 1               1  
EH2 2BY                 1               1  
EH2 2PJ                 1               1  
EH2 3AA                 3               3  
EH2 3ES                 1               1  
EH2 3NS                 1               1  
EH3 6LN                 1               1  
EH3 6NG                 1               1  
EH7 4AA                 1               1  
EH7 4AG                 1               1  
EH7 4AL                 1               1  
EH8 8DF    

In [39]:
print('There are {} unique categories.'.format(len(edinburgh_venues['Venue Category'].unique())))

There are 14 unique categories.


One hot encoding to prepare for kMeans clustering:

In [63]:
# one hot encoding
edinburgh_onehot = pd.get_dummies(edinburgh_venues[['Venue Category']], prefix="", prefix_sep="")

# add postcode column back to dataframe
edinburgh_onehot['Postcode'] = edinburgh_venues['Postcode'] 

# move postcode column to the first column
fixed_columns = [edinburgh_onehot.columns[-1]] + list(edinburgh_onehot.columns[:-1])
edinburgh_onehot = edinburgh_onehot[fixed_columns]

edinburgh_onehot.head()

Postcode  Bakery  Café  Candy Store  Creperie  Cupcake Shop  Dessert Shop  \
0  EH1 1BL       0     0            0         0             0             0   
1  EH1 1BB       1     0            0         0             0             0   
2  EH1 1BB       1     0            0         0             0             0   
3  EH1 1BB       1     0            0         0             0             0   
7  EH1 1AD       0     0            0         0             0             0   

   Gelato Shop  Gift Shop  Gourmet Shop  Ice Cream Shop  Pastry Shop  \
0            1          0             0               0            0   
1            0          0             0               0            0   
2            0          0             0               0            0   
3            0          0             0               0            0   
7            0          0             1               0            0   

   Pie Shop  Tea Room  Thai Restaurant  
0         0         0                0  
1         0         0                0  
2         0         0                0  
3         0         0                0  
7         0         0                0

In [64]:
edinburgh_grouped = edinburgh_onehot.groupby('Postcode').mean().reset_index()
edinburgh_grouped.shape

(33, 15)

To sort the venues in descending order:

In [65]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

To display the top 5 venues for each postcode:

In [66]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postcodes_venues_sorted = pd.DataFrame(columns=columns)
postcodes_venues_sorted['Postcode'] = edinburgh_grouped['Postcode']

for ind in np.arange(edinburgh_grouped.shape[0]):
    postcodes_venues_sorted.iloc[ind, 1:] = return_most_common_venues(edinburgh_grouped.iloc[ind, :], num_top_venues)

postcodes_venues_sorted.head()

Postcode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  EH1 1AD          Gourmet Shop       Thai Restaurant              Tea Room   
1  EH1 1BB                Bakery       Thai Restaurant              Tea Room   
2  EH1 1BL           Gelato Shop       Thai Restaurant              Tea Room   
3  EH1 1BQ              Tea Room       Thai Restaurant              Pie Shop   
4  EH1 1DR        Ice Cream Shop             Gift Shop                Bakery   

  4th Most Common Venue 5th Most Common Venue  
0              Pie Shop           Pastry Shop  
1              Pie Shop           Pastry Shop  
2              Pie Shop           Pastry Shop  
3           Pastry Shop        Ice Cream Shop  
4       Thai Restaurant              Tea Room

In [67]:
postcodes_venues_sorted.reset_index(drop = True, inplace = True)
postcodes_venues_sorted

Postcode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0   EH1 1AD          Gourmet Shop       Thai Restaurant              Tea Room   
1   EH1 1BB                Bakery       Thai Restaurant              Tea Room   
2   EH1 1BL           Gelato Shop       Thai Restaurant              Tea Room   
3   EH1 1BQ              Tea Room       Thai Restaurant              Pie Shop   
4   EH1 1DR        Ice Cream Shop             Gift Shop                Bakery   
5   EH1 1EZ        Ice Cream Shop       Thai Restaurant              Tea Room   
6   EH1 1HR              Pie Shop          Dessert Shop       Thai Restaurant   
7   EH1 1JQ       Thai Restaurant              Tea Room              Pie Shop   
8   EH1 1JX                  Café       Thai Restaurant              Tea Room   
9   EH1 1LS          Gourmet Shop       Thai Restaurant              Tea Room   
10  EH1 1NL                Bakery       Thai Restaurant              Tea Room   
11  EH1 1SX        Ice Cream Shop       Thai Restaurant              Tea Room   
12  EH1 2HS        Ice Cream Shop       Thai Restaurant              Tea Room   
13  EH1 2JL        Ice Cream Shop       Thai Restaurant              Tea Room   
14  EH1 2QP              Creperie       Thai Restaurant              Tea Room   
15  EH1 3AA        Ice Cream Shop       Thai Restaurant              Tea Room   
16  EH1 3DQ          Dessert Shop       Thai Restaurant              Tea Room   
17  EH1 3HP        Ice Cream Shop       Thai Restaurant              Tea Room   
18  EH1 3LD                Bakery       Thai Restaurant              Tea Room   
19  EH2 1LN          Gourmet Shop       Thai Restaurant              Tea Room   
20  EH2 2BY                Bakery       Thai Restaurant              Tea Room   
21  EH2 2PJ          Cupcake Shop       Thai Restaurant              Tea Room   
22  EH2 3AA           Pastry Shop          Dessert Shop           Candy Store   
23  EH2 3ES                Bakery       Thai Restaurant              Tea Room   
24  EH2 3NS          Dessert Shop       Thai Restaurant              Tea Room   
25  EH3 6LN                Bakery       Thai Restaurant              Tea Room   
26  EH3 6NG                Bakery       Thai Restaurant              Tea Room   
27  EH7 4AA          Gourmet Shop       Thai Restaurant              Tea Room   
28  EH7 4AG                Bakery       Thai Restaurant              Tea Room   
29  EH7 4AL        Ice Cream Shop       Thai Restaurant              Tea Room   
30  EH8 8DF        Ice Cream Shop       Thai Restaurant              Tea Room   
31  EH8 9AY        Ice Cream Shop       Thai Restaurant              Tea Room   
32  EH8 9JZ           Gelato Shop       Thai Restaurant              Tea Room   

   4th Most Common Venue 5th Most Common Venue  
0               Pie Shop           Pastry Shop  
1               Pie Shop           Pastry Shop  
2               Pie Shop           Pastry Shop  
3            Pastry Shop        Ice Cream Shop  
4        Thai Restaurant              Tea Room  
5               Pie Shop           Pastry Shop  
6               Tea Room           Pastry Shop  
7            Pastry Shop        Ice Cream Shop  
8               Pie Shop           Pastry Shop  
9               Pie Shop           Pastry Shop  
10              Pie Shop           Pastry Shop  
11              Pie Shop           Pastry Shop  
12              Pie Shop           Pastry Shop  
13              Pie Shop           Pastry Shop  
14              Pie Shop           Pastry Shop  
15              Pie Shop           Pastry Shop  
16              Pie Shop           Pastry Shop  
17              Pie Shop           Pastry Shop  
18              Pie Shop           Pastry Shop  
19              Pie Shop           Pastry Shop  
20              Pie Shop           Pastry Shop  
21              Pie Shop           Pastry Shop  
22       Thai Restaurant              Tea Room  
23              Pie Shop           Pastry Shop  
24              Pie Sho

To run the kMeans clustering algorithm using k=5 (chosen at random):

In [68]:
# set number of clusters
kclusters = 5

edinburgh_grouped_clustering = edinburgh_grouped.drop('Postcode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(edinburgh_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 2, 4, 3, 3, 1, 3, 3, 3, 0], dtype=int32)

In [69]:
# add clustering labels
postcodes_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

postcodes_edb_merged = postcodes_edb

# merge postcodes_edb with postcodes_venues_sorted to add latitude/longitude for each postcode
postcodes_edb_merged = postcodes_edb_merged.join(postcodes_venues_sorted.set_index('Postcode'), on='Postcode')

postcodes_edb_merged = postcodes_edb_merged[postcodes_edb_merged['Cluster Labels'].isnull() == False]

postcodes_edb_merged.reset_index(drop = True, inplace = True) # check the last columns!

In [86]:
postcodes_edb_merged

Postcode   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   EH1 1BL  55.950658  -3.191749             4.0           Gelato Shop   
1   EH1 1BB  55.952027  -3.189565             2.0                Bakery   
2   EH1 1AD  55.948907  -3.192608             0.0          Gourmet Shop   
3   EH1 1BQ  55.952630  -3.191329             3.0              Tea Room   
4   EH1 1DR  55.950942  -3.184711             3.0        Ice Cream Shop   
5   EH1 1EZ  55.945896  -3.189857             1.0        Ice Cream Shop   
6   EH1 1HR  55.948178  -3.187141             3.0              Pie Shop   
7   EH1 1JQ  55.948616  -3.187330             3.0       Thai Restaurant   
8   EH1 1JX  55.948143  -3.193529             3.0                  Café   
9   EH1 1LS  55.947738  -3.186150             0.0          Gourmet Shop   
10  EH1 1NL  55.949662  -3.185120             2.0                Bakery   
11  EH1 1SX  55.949627  -3.183117             1.0        Ice Cream Shop   
12  EH1 2HS  55.947070  -3.196650             1.0        Ice Cream Shop   
13  EH1 2JL  55.948629  -3.194393             1.0        Ice Cream Shop   
14  EH1 2QP  55.945674  -3.191499             3.0              Creperie   
15  EH1 3AA  55.956975  -3.185232             1.0        Ice Cream Shop   
16  EH1 3HP  55.956140  -3.189803             1.0        Ice Cream Shop   
17  EH1 3LD  55.957835  -3.188285             2.0                Bakery   
18  EH1 3DQ  55.955175  -3.191151             3.0          Dessert Shop   
19  EH2 1LN  55.953881  -3.201329             0.0          Gourmet Shop   
20  EH2 2BY  55.952937  -3.193036             2.0                Bakery   
21  EH2 2PJ  55.953117  -3.196709             3.0          Cupcake Shop   
22  EH2 3AA  55.951620  -3.201003             3.0           Pastry Shop   
23  EH2 3ES  55.952910  -3.202292             2.0                Bakery   
24  EH2 3NS  55.953862  -3.203250             3.0          Dessert Shop   
25  EH3 6LN  55.957534  -3.198062             2.0                Bakery   
26  EH3 6NG  55.961088  -3.194808             2.0                Bakery   
27  EH7 4AA  55.958465  -3.183563             0.0          Gourmet Shop   
28  EH7 4AL  55.959951  -3.183112             1.0        Ice Cream Shop   
29  EH7 4AG  55.960708  -3.181998             2.0                Bakery   
30  EH8 8DF  55.951164  -3.180282             1.0        Ice Cream Shop   
31  EH8 9AY  55.944604  -3.185926             1.0        Ice Cream Shop   
32  EH8 9JZ  55.944578  -3.188664             4.0           Gelato Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Thai Restaurant              Tea Room              Pie Shop   
1        Thai Restaurant              Tea Room              Pie Shop   
2        Thai Restaurant              Tea Room              Pie Shop   
3        Thai Restaurant              Pie Shop           Pastry Shop   
4              Gift Shop                Bakery       Thai Restaurant   
5        Thai Restaurant              Tea Room              Pie Shop   
6           Dessert Shop       Thai Restaurant              Tea Room   
7               Tea Room              Pie Shop           Pastry Shop   
8        Thai Restaurant              Tea Room              Pie Shop   
9        Thai Restaurant              Tea Room              Pie Shop   
10       Thai Restaurant              Tea Room              Pie Shop   
11       Thai Restaurant              Tea Room              Pie Shop   
12       Thai Restaurant              Tea Room              Pie Shop   
13       Thai Restaurant              Tea Room              Pie Shop   
14       Thai Restaurant              Tea Room              Pie Shop   
15       Thai Restaurant              Tea Room              Pie Shop   
16       Thai Restaurant              Tea Room              Pie Shop   
17       Thai Restaurant              Tea Room              Pie Shop   
18       Thai Restaurant              Tea Room              Pie Shop   
19       Tha

To visualise the clusters on a map of Edinburgh:

In [84]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=14)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

rainbow

['#8000ff', '#00b5eb', '#80ffb4', '#ffb360', '#ff0000']

In [85]:
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(postcodes_edb_merged['Latitude'], postcodes_edb_merged['Longitude'], postcodes_edb_merged['Postcode'], postcodes_edb_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    cluster_index = int(cluster)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster_index-1],
        fill=True,
        fill_color=rainbow[cluster_index-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters